presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -174,29 +180,44 @@ else if (!splitWaitingForAnyNode) {
         return new SplitPlacementResult(blocked, assignment);
     }
 
-    private ResettableRandomizedIterator<InternalNode> getRandomCandidates(int limit, NodeMap nodeMap, List<RemoteTask> existingTasks)
-    {
-        List<InternalNode> existingNodes = existingTasks.stream()
-                .map(remoteTask -> nodeMap.getNodesByNodeId().get(remoteTask.getNodeId()))
-                // nodes may sporadically disappear from the nodeMap if the announcement is delayed
-                .filter(Objects::nonNull)
-                .collect(toList());
-
-        int alreadySelectedNodeCount = existingNodes.size();
-        int nodeCount = nodeMap.getNodesByNodeId().size();
-
-        if (alreadySelectedNodeCount < limit && alreadySelectedNodeCount < nodeCount) {
-            List<InternalNode> moreNodes =
-                    selectNodes(limit - alreadySelectedNodeCount, randomizedNodes(nodeMap, includeCoordinator, newHashSet(existingNodes)));
-            existingNodes.addAll(moreNodes);
-        }
-        verify(existingNodes.stream().allMatch(Objects::nonNull), "existingNodes list must not contain any nulls");
-        return new ResettableRandomizedIterator<>(existingNodes);
-    }
-
     @Override
     public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, BucketNodeMap bucketNodeMap)
     {
         return selectDistributionNodes(nodeMap.get().get(), nodeTaskMap, maxSplitsPerNode, maxPendingSplitsPerTask, splits, existingTasks, bucketNodeMap);
     }
+
+    private static List<InternalNode> convertToInternalNode(NodeMap nodeMap, List<HostAddress> preferredNodes)
+    {
+        List<InternalNode> internalNodes = new ArrayList<>();
+        preferredNodes.forEach(node -> internalNodes.addAll(nodeMap.getNodesByHostAndPort().get(node)));
+        return ImmutableList.copyOf(internalNodes);
+    }
+
+    private static InternalNode chooseLeastBusyNode(List<InternalNode> candidateNodes, NodeAssignmentStats assignmentStats, OptionalInt preferredNodeCount, LimitationType limitationType, int limitation)
```

Comment:
- Instead of having `assignmentStats` and `LimitationType`, what you need is `Function<InternalNode, Integer> splitCountProvider`
- nit: `limitation` -> `maxSplitCount`